packages/flutter/lib/src/rendering/object.dart

```diff
@@ -1369,13 +1369,50 @@ class PipelineOwner with DiagnosticableTreeMixin {
       return true;
     }());
     try {
+      // This has to be top-to-down order since the geometries of a child and its
+      // subtree depends on ancestors' transforms and clips. If it updates child
+      // first, it may use dirty geometry in parent's semantics node to
+      // calculate the geometries in the subtree.
       final List<RenderObject> nodesToProcess =
-          _nodesNeedingSemantics.toList()
+          _nodesNeedingSemantics
+              .where((RenderObject object) => !object._needsLayout && object.owner == this)
+              .toList()
             ..sort((RenderObject a, RenderObject b) => a.depth - b.depth);
       _nodesNeedingSemantics.clear();
-      for (final RenderObject node in nodesToProcess) {
-        if (node._needsSemanticsUpdate && node.owner == this) {
-          node._updateSemantics();
+      if (nodesToProcess.isNotEmpty) {
+        for (final RenderObject node in nodesToProcess) {
+          if (!kReleaseMode) {
+            FlutterTimeline.startSync('Semantics.GetFragment');
+          }
+          if (node._semantics.parentDataDirty) {
+            // This is a node that are blocked by sibling. Otherwise, the
+            // parent node this node would have update this node's parent data.
+            continue;
+          }
+          node._semantics.updateChildren();
```

Comment:
I left a comment about this elsewhere, but how do we avoid building the same subtree twice? It seems like `updateChildren` recurses to build the whole subtree if necessary, but some of these nodes may also be part of the `nodesToProcess` and have `updateChildren` called on them again.